src/data/polynomial/iterated_deriv.lean

```diff
@@ -0,0 +1,172 @@
+/-
+Copyright (c) 2020 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+
+import data.polynomial.derivative
+import logic.function.iterate
+import tactic.ring tactic.linarith
+
+/-!
+# Theory of iterated derivative
+We define and prove some lemmas about iterated (formal) derivative for polynomials over a semiring.
+-/
+
+noncomputable theory
+
+open finset nat polynomial
+open_locale big_operators
+
+namespace polynomial
+universes u
+variable {R : Type u}
+
+section semiring
+
+variables [semiring R] (f p q: polynomial R) (n k : ℕ)
+
+/--Defining the n-th derivative of polynomial f-/
+def iterated_deriv : polynomial R := derivative ^[n] f
```

Comment:
It would be better to define `iterated_deriv : polynomial R →ₐ[R] polynomial R`, so we get the properties of `alg_hom`s for free. (You might need to import `ring_theory.algebra` first.)

Reply:
But `R` in general is not commutative. When I try to define it as an algebra homomorphism, lean complains `R` might not be a `comm_semiring`

Reply:
Aha, of course. Then could you manually add lemmas for `iterated_deriv` of `C x`, `0`, `1`, `X`, `-p`, `p - q` and `r • p`? That should be enough to cover most of the polynomial constructions.

Reply:
please see the new commit